### DISCUSSION TITLE
Trying to re-initialize database via ORM in a Jupyter Notebook

### DISCUSSION THREAD

**Discussion by jat255** [sqlalchemy/sqlalchemy#7915] [Usage Questions] [answered]
I'm trying to use the ORM (SQLAlchemy 1.4.32) to demonstrate some database operations, and I'm running into a hiccup trying to re-initialize my database (SQLite) when I re-run a cell. I'll note that it all works if I completely restart my Jupyter kernel (which is a fine workaround), but I'm trying why it doesn't work in the general case as I'm learning more about SQLAlchemy. I have the following files defining my model in a folder called `db_demo`:

<details>
  <summary>base.py</summary>

 ```python
from sqlalchemy import create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker
import os

Base = declarative_base()

db_path = os.environ.get('DB_PATH', None)
db_echo = os.environ.get('DB_ECHO', 'False').lower() == 'true'
engine = create_engine(f"sqlite:///{db_path}",
                       echo=db_echo, future=True)

Session = sessionmaker(bind=engine)
Base = declarative_base()
metadata = Base.metadata
```

</details> 

<details>
  <summary>model.py</summary>

 ```python
from .base import Base

class DateType(Base):
    __tablename__ = "date_types"
    ...

class RelationType(Base):
    __tablename__ = "relation_types"
    ...

# etc.

# some of the above classes represent controlled vocabularies, so I define those objects in model.py as well
# to be added as rows when the DB is initialized:

dt1 = DateType(date_type='dt1', definition='def 1')
dt2 = DateType(date_type='dt2', definition='def 2')
controlled_vals = [dt1, dt2]
# etc. ...
```

</details> 

In my notebook, the following works fine when first starting the kernel:

```python
import os
import pathlib

# setup environment for database config
os.environ['DB_PATH'] = str(pathlib.Path('.') / 'my.db')  # path to DB file to use
os.environ['DB_ECHO'] = 'true'   # whether to have SQLAlchemy echo SQL to the console
os.environ['DB_FRESH'] = 'true'   # whether to delete any existing data in database

# import database ORM definitions
from db_demo.base import Session, Base, metadata, engine
from db_demo import model

# drop all data/tables in DB
if os.environ.get('DB_FRESH', 'true').lower() == 'true':
    metadata.drop_all(engine)
    
metadata.create_all(engine)
    
# import controlled vocabulary definitions
with Session() as session:
    session.add_all(model.controlled_vals)
    session.commit()
```

As expected, `my.db` at this point contains all my tables and the controlled vocabulary like I expect. If I run this cell again however, all the tables are dropped (`metadata.drop_all(engine)`) and recreated (`metadata.create_all(engine)`), but the controlled vocabulary rows are not added again (in `session.add_all(model.controlled_vals)`) -- no INSERT SQL is emitted when re-running the cell a second time.

I _assume_ (not sure) that this is because as far as `Session` is concerned, those objects have not changed and since they've already been added to the database, they do not need to be re-emitted. I tried various combinations of `expire()`, `expunge()`, etc., but could not get it to work. 

Any ideas what I'm missing, or is this usage pattern not supported/recommended? I'm mostly curious _why_ it's not working the way I expect, since that will help me understand the ORM/sessions better.

**Comment by zzzeek** [comment, marked as answer]
I wasnt able to see what the issue was but yes, if you are rerunning the "with Session()" part, but then persisting the same instances (object references) that you already persisted before, those objects have mutated.   The objects you pass to session.add() are state tracking, and they have tracked that they were persisted. 

the easy solution is to create the test data set each time you need it:

```py
def create_test_data():
    dt1 = DateType(date_type='dt1', definition='def 1')
    dt2 = DateType(date_type='dt2', definition='def 2')
    controlled_vals = [dt1, dt2]

    return {
       "dt1": dt1,
       "dt2": dt2,
       "controlled_vals": controlled_vals
    }
```

or something like that, where each time your "cell" runs you create new objects to work with.

**Reply by jat255** [reply]
Thank you! That makes a lot of sense and is working now with a function inside of `model.py` (I called `get_controlled_vocab()` that returns a dictionary of all the instances. That way they're new each time I run the cell. Thanks for your time and help.
